bin/export-codePost-grades

```diff
@@ -0,0 +1,254 @@
+                          "~/codepost-config.yaml", "~/.codepost-config.yaml"]
+
+    location = None
+    for p in possible_locations:
+        path = os.path.abspath(os.path.expanduser(p))
+        if os.path.exists(path):
+            location = path
+            break
+
+    if location == None:
+        _print_err(
+            "No codepost-config.yaml configuration file detected. Exiting.",
+            fatal=8)
+    else:
+        _print_info("Configuration path: {}".format(location))
+
+    config = load(open(location), Loader=Loader)
+    _print_info("Configuration content: {}".format(config))
+
+    return params, config
+
+
+clparams, config = process_command_line(parser)
+
+grades = cP.get_course_grades(
+    api_key=config["api_key"],
+    course_name=config["course_name"],
+    course_period=config["course_period"])
+
+# JSON output
+if clparams["json"]:
+    import json
+    if clparams["pretty"]:
+        print(json.dumps(grades, indent=2))
+    else:
+        print(json.dumps(grades))
+    sys.exit(0)
+
+
+course = cP.get_course_roster_by_name(
+    api_key=config["api_key"],
+    course_name=config["course_name"],
+    course_period=config["course_period"])
+
+# CSV output
+
+# Figure out headers
+headers = clparams.get("a", list())
+if headers == None or len(headers) == 0:
+    headers = []
+    for aid in course["assignments"]:
+        assignment_info = cP.get_assignment_info_by_id(
+            api_key=config["api_key"],
+            assignment_id=aid)
+        headers.append(assignment_info["name"])
+
+    if not clparams["show_empty"]:
+        used_assignments = set()
+        for student_grades in grades.values():
+            used_assignments = used_assignments.union(
+                set(student_grades.keys()))
+
+        headers = [header for header in headers if header in used_assignments]
+
+# Figure out students
+students = clparams.get("s", list())
+if len(students) > 0:
+    # Normalize input
+    students = map(lambda n: config["user_pattern"].format(n), students)
+
+else:
+    students = course.get("students", list())
+
+    if clparams["include_inactive"]:
+        students += course.get("inactive_students", list())
+
+csv_lines = []
+
+header_line = None
+
+if clparams["blackboard"]:
+    if not "lms_ids" in config:
+        _print_warn(
+            "Blackboard column IDs requested but not available in configuration")
+    else:
+        lms_ids = config["lms_ids"]
+        lms_format = config.get("lms_id", '"{name} | {id}"')
+
+        def format_header_title(title):
+            lms_id = lms_ids.get(title, "")
+            return lms_format.format(
+                name=title, id=lms_id)
+
+        header_line = ",".join(map(format_header_title, headers))
+
+if header_line == None:
+    header_line = ",".join(map('"{}"'.format, headers))
+
+csv_lines.append('"Username",{}'.format(header_line))
+
+students.sort()
+
+fmt_string = ",".join(["{}"] * len(header_line.split(",")))
```

Comment:
`len(header_line.split(","))` should be `(len(header_line.split(",")) + 1)` in order to include the last column (we need to count the Username column)

Reply:
Very good catch! Since I remember thinking specifically that I got this one right, I have added a comment for future tinkerers, since I think my coding style might here have made things more complex than necessary:

```python
# A "{},{},{},{},{},{}..." formatting string, which is
# equivalent to plain concatenation of all columns, separated
# by commas. The len() + 1 comes from the "Username" field
# which is not part of the header_line.

fmt_string = ",".join(["{}"] * (len(header_line.split(",")) + 1))
```